### PR TITLE
Include substitutions in notification

### LINKED ISSUE #16: When notifying people, include their substitutions

### PR DESCRIPTION
**Pull request by zUnixorn**
closes #16

### CODE REVIEW

File: src/substitution_schedule.rs [hunk 1]
```diff
@@ -11,40 +11,59 @@ use lopdf::Document;
...
 
 use crate::tabula_json_parser::parse;
+use std::ops::Range;
 
 #[derive(Serialize, Deserialize, PartialOrd, PartialEq)]
 pub struct Substitutions {
 	#[serde(rename(serialize = "0"))]
 	#[serde(rename(deserialize = "0"))]
-	pub block_0: String,
+	#[serde(skip_serializing_if = "Option::is_none")]
```
Comment (zUnixorn) [inline src/substitution_schedule.rs:20]: Deserialization could be affected

File: src/main.rs [hunk 2]
```diff
@@ -130,21 +130,14 @@ async fn check_weekday_pdf(day: Weekdays, pdf_getter: Arc<SubstitutionPDFGetter<
...
-				if let Some(old_substitutions) = old_schedule.get_substitutions(class.as_str()) {
-					if new_substitutions != old_substitutions {
-						discord.notify_users_for_class(class.as_str(), day).await?;
-					}
-				}
-			} else {
-				discord.notify_users_for_class(class.as_str(), day).await?;
-			}
+	if let Some(old_schedule) = old_schedule_option {
+		if !new_schedule.get_entries().keys().all(|k| old_schedule.get_entries().contains_key(k)) {
```
Comment (flammenderdrache) [inline src/main.rs:134]: No comparison wether there was actually a change in schedule (Not comparing old != new)

File: src/discord.rs [hunk 3]
```diff
@@ -111,6 +113,25 @@ impl ClassesAndUsers {
 		}
 		classes
 	}
+
+	pub fn to_inside_out(&self) -> HashMap<u64, HashSet<&String>> {
+		let mut inside_out: HashMap<u64, HashSet<&String>> = HashMap::new();
+
+		for (class, users) in &self.classes_and_users {
+			for user in users {
+				if let Some(mut classes) = inside_out.get_mut(&user) {
```
Comment (flammenderdrache) [inline src/discord.rs:122]: maybe use the same method we use in insert_user()

File: src/substitution_schedule.rs [hunk 4]
```diff
@@ -81,7 +100,7 @@ impl SubstitutionSchedule {
 				for (i, substitution_part) in table[row][1..].iter().enumerate() {
 					let substitutions = entries.get_mut(&classes[i]).unwrap();
 
-					let block = match lesson_idx {
+					let block_opt = match lesson_idx {
```
Comment (flammenderdrache) [inline src/substitution_schedule.rs:103]: opt or option? consistency please :P

File: src/substitution_schedule.rs [hunk 5]
```diff
@@ -161,6 +180,18 @@ impl SubstitutionSchedule {
 	pub fn get_substitutions(&self, class: &str) -> Option<&Substitutions> {
 		self.entries.get(class)
 	}
+
+	pub fn get_entries(&self) -> &HashMap<String, Substitutions> { &self.entries }
+
+	pub fn get_entries_portion(&self, classes: &HashSet<&String>) -> HashMap<String, &Substitutions> {
```
Comment (flammenderdrache) [inline src/substitution_schedule.rs:186]: What we talked about. Skip non existant classes instead of unwrapping

**Review by flammenderdrache — CHANGES_REQUESTED** [review]
Please change the commented things :)

File: src/substitution_pdf_getter.rs [hunk 6]
```diff
@@ -44,7 +44,7 @@ impl Display for Weekdays {
 
 impl From<Weekday> for Weekdays {
 	fn from(day: Weekday) -> Self {
-		match day {
+match day {
```
Comment (FlareFlo) [inline src/substitution_pdf_getter.rs:47]: Failed indent?

**Review by flammenderdrache — APPROVED** [review]

**Pull request merged**